x-pack/plugins/security_solution/server/lib/detection_engine/signals/build_bulk_body.ts

```diff
@@ -21,18 +22,27 @@ import { buildEventTypeSignal } from './build_event_type_signal';
 import { EqlSequence } from '../../../../common/detection_engine/types';
 import { generateSignalId, wrapBuildingBlocks, wrapSignal } from './utils';
 
-// format search_after result for signals index.
+/**
+ * Formats the search_after result for insertion into the signals index. We first create a
+ * "best effort" merged "fields" with the "_source" object, then build the signal object,
+ * then the event object, and finally we strip away any additional temporary data that was added
+ * such as the "threshold_result".
+ * @param ruleSO The rule saved object to build overrides
+ * @param doc The SignalSourceHit with "_source", "fields", and additional data such as "threshold_result"
+ * @returns The body that can be added to a bulk call for inserting the signal.
+ */
 export const buildBulkBody = (
   ruleSO: SavedObject<AlertAttributes>,
   doc: SignalSourceHit
 ): SignalHit => {
-  const rule = buildRuleWithOverrides(ruleSO, doc._source!);
+  const mergedDoc = mergeMissingFieldsWithSource({ doc });
+  const rule = buildRuleWithOverrides(ruleSO, mergedDoc._source ?? {});
   const signal: Signal = {
-    ...buildSignal([doc], rule),
-    ...additionalSignalFields(doc),
+    ...buildSignal([mergedDoc], rule),
+    ...additionalSignalFields(mergedDoc),
```

Comment:
Understand sending the merged doc down the chain here to maintain existing compatibility of anything using `_source`, but could be interesting with any additional fields coming from `fields`.